tools/owl-vision/src/search.ts

```diff
@@ -0,0 +1,255 @@
+		if (result) {
+			return result;
+		}
+
+		vscode.window.showWarningMessage(`No template found for "${templateName}"`);
+		return undefined;
+	}
+
+	private async findComponentFromTemplateName(templateName: string): Promise<SearchResult | undefined> {
+		const result = await this.find(templateName, `template\\s*=\\s*["'](${templateName})["']`, this.componentCache, "js");
+		if (result) {
+			return result;
+		}
+
+		vscode.window.showWarningMessage(`Could not find a component for "${templateName}"`);
+		return undefined;
+	}
+
+	private getTemplateNameInJS(str: string): string | undefined {
+		const matches = new RegExp(`template\\s*=\\s*["']([a-zA-Z0-9_\\-\\.]+)["']`, 'g').exec(str);
+		if (matches) {
+			return matches[1];
+		} else {
+			return undefined;
+		}
+	}
+
+	private getTemplateNameInXML(str: string): string | undefined {
+		const matches = new RegExp(`t-name="([a-zA-Z0-9_\\-\\.]+)"`, 'g').exec(str);
+		if (matches) {
+			return matches[1];
+		} else {
+			return undefined;
+		}
+	}
+
+	private async find(
+		name: string,
+		searchQuery: string,
+		cache: Map<string, SearchResult>,
+		fileType: "js" | "xml",
+	) {
+		if (cache.has(name)) {
+			return cache.get(name);
+		}
+
+		const include = `{${vscode.workspace.getConfiguration().get(`owl-vision.${fileType}.include`)}}`;
+		const exclude = `{${vscode.workspace.getConfiguration().get(`owl-vision.${fileType}.exclude`)}}`;
+		const files = await this.getFiles(name, include, exclude);
+
+		for (const file of files) {
+			let result = await this.findInFile(file, new RegExp(searchQuery, 'g'), 6);
+			if (result) {
+				cache.set(name, result);
+				return result;
+			}
+		}
+		return undefined;
+	}
+
+	private async getFiles(
+		searchQuery: string,
+		include: vscode.GlobPattern,
+		exclude: vscode.GlobPattern,
+	): Promise<Array<vscode.Uri>> {
+		const files = await vscode.workspace.findFiles(include, exclude);
+		const parts = searchQuery.split(".").flatMap(s => s.split(/\.?(?=[A-Z])/)).map(s => s.toLowerCase());
+		const currentDir = this.currentDocument ? path.dirname(this.currentDocument.uri.path) : "";
+
+		const results = files.map(file => {
+			const filepath = file.path.toLowerCase();
+			let score = 0;
+			if (path.dirname(filepath) === currentDir) {
+				score += 99;
+			}
+			for (const part of parts) {
+				if (filepath.includes(part)) {
+					score++;
+				}
+			}
+			if (filepath.includes("legacy")) {
+				score = score / 2.0;
+			}
+			return { score, file };
+		}).sort((a, b) => {
+			return a.score > b.score ? -1 : 1;
+		}).slice(0, 50);
+
+		return results.map(r => r.file);
+	}
+
+	private async findInFile(
+		file: vscode.Uri,
+		searchQuery: RegExp,
+		selectionStartOffset: number,
+	): Promise<SearchResult | undefined> {
+		const document = await vscode.workspace.openTextDocument(file);
+		const text = document.getText();
+		const regex = searchQuery;
+		const match = regex.exec(text);
```

Comment:
`RegExp.prototype.exec` is a footgun because it mutates the regexp object (which this function doesn't own), prefer `String.prototype.match(All)` (you'll need matchAll if your regexp are global and you want the indices of the match) or `String.prototype.search` instead